Update Identity custom user data documentation for .NET 8/9/10 with MVC instructions and troubleshooting (#36222)
dotnet/AspNetCore.Docs: aspnetcore/security/authentication/add-user-data.md

---
title: Add, download, and delete user data to Identity in an ASP.NET Core project
ai-usage: ai-assisted
author: wadepickett
description: Learn how to add custom user data to Identity in an ASP.NET Core project. Delete data per GDPR.
ms.author: wpickett
ms.custom: mvc

> ms.date: 04/02/2026

Reviewer: Should this be 07/02?

Reviewer 2: @Tdykstra, resolved in this last commit, thanks.

> ms.date: 04/02/2026

Reviewer 2: ```suggestion
ms.date: 07/05/2026
```

uid: security/authentication/add-user-data
---
# Add, download, and delete custom user data to Identity in an ASP.NET Core project

This article shows how to:

* Add custom user data to an ASP.NET Core web app.
* Mark the custom user data model with the <xref:Microsoft.AspNetCore.Identity.PersonalDataAttribute> attribute so it's automatically available for download and deletion. Making the data able to be downloaded and deleted helps meet [GDPR](xref:security/gdpr) requirements.

The project sample is created from a Razor Pages web app, but the instructions are similar for an ASP.NET Core MVC web app.

[View or download sample code](https://github.com/dotnet/AspNetCore.Docs/tree/main/aspnetcore/security/authentication/add-user-data) ([how to download](xref:fundamentals/index#how-to-download-a-sample))

## Prerequisites

:::moniker range=">= aspnetcore-10.0"

[!INCLUDE [](~/includes/10.0-SDK.md)]

> ## Quick start: Add two custom properties

Reviewer: I think we typically treat quickstart as one word.

Reviewer 2: @Tdykstra, resolved in this last commit, thanks.

> ## Quick start: Add two custom properties

Reviewer 2: ```suggestion
## Quickstart: Add two custom properties
```

For the common case of adding two custom properties such as `FirstName` and `LastName`:

1. Create a new ASP.NET Core project with Individual Accounts authentication
1. Scaffold Identity to override the Register and Manage/Index pages
1. Add custom properties such as `FirstName` and `LastName` to the user class, with the `[PersonalData]` attribute.
1. Update the Register and Manage/Index pages to include input fields for your properties
1. Create and apply a migration

> 1. Test registration and profile management

Reviewer: ```suggestion
1. Update the Register and Manage/Index pages to include input fields for your properties.
1. Create and apply a migration.
1. Test registration and profile management.
```

## Prerequisites

:::moniker range=">= aspnetcore-10.0"

[!INCLUDE [](~/includes/10.0-SDK.md)]

## Quick start: Add two custom properties

For the common case of adding two custom properties such as `FirstName` and `LastName`:

1. Create a new ASP.NET Core project with Individual Accounts authentication
1. Scaffold Identity to override the Register and Manage/Index pages

> 1. Add your custom properties to the user class (e.g., `FirstName`, `LastName`) with the `[PersonalData]` attribute

Reviewer 2: ```suggestion
1. Add your custom properties to the user class, for example `FirstName`, `LastName`, with the `[PersonalData]` attribute
```

  * Test the **Delete** button, which deletes the logged on user.

## Add custom user data to the Identity DB

Update the `IdentityUser` derived class with custom properties. If you named the project WebApp1, the file is named `Areas/Identity/Data/WebApp1User.cs`. Update the file with the following code:

[!code-csharp[](add-user-data/samples/10.x/SampleApp/Areas/Identity/Data/WebApp1User.cs)]

Properties with the [PersonalData](xref:Microsoft.AspNetCore.Identity.PersonalDataAttribute) attribute are:

* Deleted when the `Areas/Identity/Pages/Account/Manage/DeletePersonalData.cshtml` Razor Page calls `UserManager.Delete`.
* Included in the downloaded data by the `Areas/Identity/Pages/Account/Manage/DownloadPersonalData.cshtml` Razor Page.

### Update the `Account/Manage/Index.cshtml` page

Update the `InputModel` in `Areas/Identity/Pages/Account/Manage/Index.cshtml.cs` with the following highlighted code:

[!code-csharp[](add-user-data/samples/10.x/SampleApp/Areas/Identity/Pages/Account/Manage/Index.cshtml.cs?name=snippet&highlight=30-38,54-55,97-107)]

Update the `Areas/Identity/Pages/Account/Manage/Index.cshtml` with the following highlighted markup:

> [!code-cshtml[](add-user-data/samples/10.x/SampleApp/Areas/Identity/Pages/Account/Manage/Index.cshtml?highlight=18-25,81-82)]

Reviewer 2: ```suggestion
[!code-cshtml[](add-user-data/samples/10.x/SampleApp/Areas/Identity/Pages/Account/Manage/Index.cshtml?highlight=18-25)]
```
